syntax = "proto2";
package addressbook;

message Person {
  required string name = 1;
  required sint64 id = 2;
  optional string email = 3;

  enum PhoneType {
    MOBILE = 0;
    HOME = 1;
    WORK = 2;
  }

  message PhoneNumber {
    required string number = 1;
    optional PhoneType type = 2;
  }

  repeated PhoneNumber phones = 4;
  required string last_updated = 5;
}

message AddressBookGetQuery {
  required string name = 1;
}

message AddressBookDeleteQuery {
  required sint64 id = 1;
}

message Ok {
  required string msg = 1;
}

service AddressBook {
  rpc Get (AddressBookGetQuery) returns (stream Person) {}
  rpc Update (Person) returns (Ok) {}
  rpc Set (Person) returns (Ok) {}
  rpc Delete (AddressBookDeleteQuery) returns (Ok) {}
}

